src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs

```diff
@@ -0,0 +1,187 @@
+        {
+            throw new SerializationException(SR.Format(SR.Serialization_InvalidTypeOrAssemblyName, rawName, libraryRecord.RawLibraryName));
+        }
+
+        if (typeName.AssemblyName is null)
+        {
+            // Sample invalid input that could lead us here:
+            // TypeName: System.Collections.Generic.List`1[[System.String
+            // LibraryName: 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]
+            // Since the flag is ON, we know it's mangling and we provide missing information.
+            typeName = typeName.WithCoreLibAssemblyName();
+        }
+
+        return typeName;
+    }
+
+    internal static TypeName ParseSystemRecordTypeName(this string rawName, PayloadOptions payloadOptions)
+        => ParseWithoutAssemblyName(rawName, payloadOptions)
+                .WithCoreLibAssemblyName(); // We know it's a System Record, so we set the LibraryName to CoreLib
+
+    internal static TypeName WithCoreLibAssemblyName(this TypeName systemType)
+        => systemType.WithAssemblyName(s_CoreLibAssemblyName ??= AssemblyNameInfo.Parse("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089".AsSpan()));
+
+    private static TypeName WithAssemblyName(this TypeName typeName, AssemblyNameInfo assemblyName)
+    {
+        if (!typeName.IsSimple)
+        {
+            if (typeName.IsArray)
+            {
+                TypeName newElementType = typeName.GetElementType().WithAssemblyName(assemblyName);
+
+                return typeName.IsSZArray
+                    ? newElementType.MakeSZArrayTypeName()
+                    : newElementType.MakeArrayTypeName(typeName.GetArrayRank());
+            }
+            else if (typeName.IsConstructedGenericType)
+            {
+                TypeName newGenericTypeDefinition = typeName.GetGenericTypeDefinition().WithAssemblyName(assemblyName);
+
+                // We don't change the assembly name of generic arguments on purpose.
+                return newGenericTypeDefinition.MakeGenericTypeName(typeName.GetGenericArguments());
+            }
+            else
+            {
+                // BinaryFormatter can not serialize pointers or references.
+                ThrowHelper.ThrowInvalidTypeName(typeName.FullName);
+            }
+        }
+
+        return typeName.MakeSimpleTypeName(assemblyName);
+    }
+
+    private static TypeName ParseWithoutAssemblyName(string rawName, PayloadOptions payloadOptions)
+    {
+        if (!TypeName.TryParse(rawName.AsSpan(), out TypeName? typeName, payloadOptions.TypeNameParseOptions)
+            || typeName.AssemblyName is not null) // the type and library names should be provided separately
+        {
+            throw new SerializationException(SR.Format(SR.Serialization_InvalidTypeName, rawName));
+        }
+
+        return typeName;
+    }
+
+    private static ArraySegment<char> GetAssemblyQualifiedName(string typeName, string libraryName)
```

Comment:
Consider renaming to `RentAssemblyQualifiedName`, so it's clearer from context that the rent-return is the responsibility of the caller.